### ISSUE TITLE
Struggling to 'Update Links' // Code taken from 2nd demo

### ISSUE THREAD

**Issue by harrydry** [luruke/barba.js#137]
Hi, 

Firstly, massive genuine thanks a lot for Barba. Judging by the examples it is absolutely perfect for my site... I just can't get it to do what I want at the minute!!!

I basically want to replicate this demo: http://barbajs.org/demo/nextprev/index.html, except: 

On the index.html not have an option to click PREV  & 
On the 4.html not have an option to click NEXT

So I have taken the source code from the above demo into my text-editor and am just trying to make it work. 

I'm guessing this is still very simple. But there is some stuff I'm confused by: 

My theory was that if I removed:

```
<a class="nav prev" href="4.html">PREV</a>
```

on the index.html page, maybe the prev button would go and I could cycle through as I want to. 

But on removal of this code, if I click the 'next' button, the page refreshes but nothing happens. 
What is particularly interesting is that the url changes to 2.html despite the image not changing. 

I understand this a little bit as perhaps the promise can no longer run so the code fails. 
___

So then I believed that if i added the class:  'disabled' like such: 

```
<a class="nav prev disabled" href="4.html">PREV</a>
```
and added in the css 

```
.disabled {
    pointer-events: none;
    position: relative;
    opacity: .20;
} 
```
but even if I add .disabled to just the index.html 
the prev button remains disabled for all of 2.html, 3.html, 4.html despite the  class .disabled is not added to them.  

I'm certain that these problems are to do with my lacking of understanding of how the 'links update' in the demo example: 

```
data-prev="4.html" and data-next="2.html"

    var PrevLink = document.querySelector('a.prev');
    var NextLink = document.querySelector('a.next');

this.updateLinks();
        updateLinks: function() {
            PrevLink.href = this.newContainer.dataset.prev;
            NextLink.href = this.newContainer.dataset.next;
        }, 
```

In conclusion: 

I mean, this stuff is slightly beyond me (but only slightly), I understand Tweens and most of the other stuff, just stumped by the 'update links' thing. 

i noticed that on your third demo (the circles one) you wrote all the update links in the JS: 

      var oldText = this.oldContainer.querySelector('.text');
      var newText = this.newContainer.querySelector('.text');
      var newLinkPrev = this.newContainer.querySelector('.nav.prev');
      var newLinkNext = this.newContainer.querySelector('.nav.next');

Is there a good tutorial / knowledge source to understand whats going on to 'update the links' so to speak? Or is there an alternative option which is well documented? 

I'm guessing it is a quick job to do what I'm trying to do? 

Very sorry for the long message. im spreading the barba word to my friends.
Genuinely a massive thanks  for any help!!

**Comment by harrydry** [comment]
I'm aware I throw a lot at the wall here. 
Take any direction with your response
Perhaps I need to familiarise myself with another language?

Take some smaller steps before I can understand it?

**Comment by trimacdo** [comment]
```
<nav class="navigation-links hidden" role="navigation">
                <ul class="navigation-links__ul">
                    <li><a class="navigation-links--active" href="./home.html">THE FIRM</a>
                        <ul class="navigation-links__ul-inner-ul">
                            <li><a href="./capabilities.html">CAPABILITIES</a></li>
                            <li><a href="./approach.html">APPROACH</a></li>
                            <li><a href="./team.html">MEET THE TEAM</a></li>
                        </ul>
                    </li>
                    <li><a href="./ourwork.html">OUR WORK</a>

                        <ul class="navigation-links__ul-inner-ul">
                            <li><a href="./inprogress.html">INPROGRESS</a></li>
                            <li><a href="./completed.html">COMPLETED</a></li>
                            <li><a href="./landdevelopments.html">DEVELOPMENTS</a></li>
                        </ul></li>
                    <li><a href="./contact.html">CONTACT</a></li>
                </ul>
            </nav>

```
```
Barba.Dispatcher.on('linkClicked', function (HTMLElement) {
  // On Link click, get all the elements with --active modifier
  // then remove them, as well close the mobile menu by adding
  // class "hidden"
  const activeNavElements = document.getElementsByClassName('navigation-links--active');
  let i;
  for (i = 0; i < activeNavElements.length; i += 1) {
    activeNavElements[i].classList.remove('navigation-links--active');
  }
  const mobileMenu = document.querySelector('.navigation-links');
  mobileMenu.classList.add('hidden');

  // Get HTML Elements href pointer, then find the navigation
  // link with the matching pointer and add the --active class
  // modifier
  const elementLink = HTMLElement.getAttribute('href');
  const nav = document.querySelector('.navigation-links');
  const navLinkToAddActive = nav.querySelector(`a[href="${elementLink}"]`);
  navLinkToAddActive.classList.add('navigation-links--active');
});
```
This is a code snippet I used in my project.

The important parts are:

```
 const elementLink = HTMLElement.getAttribute('href');
  const nav = document.querySelector('.navigation-links');
  const navLinkToAddActive = nav.querySelector(`a[href="${elementLink}"]`);
  navLinkToAddActive.classList.add('navigation-links--active');

```

As you can see here, I get the href value of the element clicked on, then get the same element that has an href of the href value of the element clicked on, we then add the --active class to that element. I wrote this because I had buttons within my barba container, but my navigation is a static element sitting outside of barba container, so it doesn't get updated on page transition, so you have to manually update the active tag with javascript.

This same logic can be applied in your problem.

I will just use pseudo code here, as it is very basic javascript. 

If your HTMLElement's href value matches the last page, that you wish to disable the next button on, then add the class disabled. & vice versa, if the href value doesn't match last page, make sure disabled class is removed from next button.

If your HTML element's href value matches the first page, add the class disabled to the prev button element. & vice versa, if the href value doesn't match first page, make sure disabled class is removed.

In your base html, add disabled class to the first page in the sequence (so if your user lands on that page to begin with, it's already disabled) and the same for the last page, add the disabled class to the next button on the last page.

I hope this helps.

**Comment by harrydry** [comment]
will try and implement this later tonight. 

massive thanks for taking the time to send this.

**Comment by harrydry** [comment]
Hi Tristan, 

Ok, some of your stuff is making sense (just been working through and trying to add it). 
Unfortunately, things are very busy or me at the minute (school exams) so havn't managed to finish it. 

Is there any chance you could link me to your website where this animation happens? 
Or upload the project files? I feel watching it happen and inspecting etc... will help me piece together. 

I'm beggining to click , but perhaps I overestimate my JS knowledge. 

No worries if not, I'll upload an 'attempt' based off of the demo when I get some time. 
In the mean-time my problem is found here: 

http://barbajs.org/demo/nextprev/3.html

To summarise: 

How does this work with 'prev' disabled on first page and 'next' disabled on 4.html. 

massive thanks again, Tristan

**Comment by trimacdo** [comment]
https://cartoonist-hamster-11141.netlify.com/templates/home.html

You can see it in action when you click on any of the navigation links, or any of the buttons that link to another page. 

Unfortunately my code is minified and will be quite hard to discern anything from in this demo. The code I have provided you above is all you will need. 

Could you throw up a demo, or your code you're trying to setup and I can walk you through it?

In the mean time I'll try to throw up a working example myself.

**Comment by trimacdo** [comment]
Barba.Dispatcher.on('linkClicked', function (HTMLElement) {
  // Get HTML Elements href pointer, then find the navigation
  // link with the matching pointer and add the --active class
  // modifier
  const elementLink = HTMLElement.getAttribute('href');
  const nextButton = document.querySelector('.next');
  const prevButton = document.querySelector('.prev');

  if (elementLink === 'firstpage.html' ) {
     prevButton.classList.add('disabled');
  }

 if (elementLink === 'lastpage.html' ) {
     nextButton.classList.add('disabled');
  }

 if (nextButton.classList.contains('disabled') && elementLink != 'lastpage.html' ) {
   nextButton.classList.remove('disabled');
 }

 if (prevButton.classList.contains('disabled') && elementLink != 'firstpage.html' ) {
   prevButton.classList.remove('disabled');
 }
});


I didn't test this or anything, but this is how I would initially do it, I didn't dive too deep into this so it may not be the best way, but some along those lines should do it.

PSEUDO CODE EXPLANATION:

On click of an eligible PJAX (Barba) link, then we will run the function, with an argument of HTMLElement, where HTMLElement is the link that was clicked on.

We then get the href value of that element clicked on (figuring out what page we are going to)

then we have some if statements. 

If the href value equals the FIRST page, we will disable the prev button.

if the href value equals the LAST page, we will disable the next button.

if the nextButton contains a class of disabled AND we're NOT going to the last page, we can remove the disabled class.

if the prevButton contains a class of disabled AND we're not going to the first page, we can remove the disabled class.



I think that covers everything you were asking?

**Comment by harrydry** [comment]
1. www.theconsumers.co  

The website where im trying to get the animation effect.
You click the down arrow and it should scroll down to the next page of the book using the 'barba.js'. You click the up arrow and it scrolls back up. but obviously i want the up-arrow disabled for the first page. and the down arrow disabled on the last page of the book. 
 
(note there is a typing effect on the first page which can be skipped by clicking the fast-forward button. Also only a few pages uploaded as of yet. ) 

2. www.thebooktheworldwrote.com 

the simple demo which i have been basing my ideas from. (luke's demo) 

______

Will read through your words tomorrrow and try and piece it together (late in the UK) 
really am very grateful. my first ever git-hub post and didn't expect this kinda help.

**Comment by trimacdo** [comment]
www.theconsumers.co 

in this demo, your links that are controlling which page you're going are are WITHIN the barba container.

This means that when you click on the link, its removing the container that contains your next/prev button links, and then adding in the new container, WHICH HOLDS NEW NEXT PREV BUTTONS.

You'll have to nest the next/prev buttons OUTSIDE of the element that has the barba-container class.

Then the code I wrote above with the if statements, you can modify to fit your urls and make sure theres no errors because I didn't write it in a text editor.

**Comment by harrydry** [comment]
great explantion there. many thanks 

getting there - www.theconsumers.co. still some mistakes from me. 

So the 'disabled' class does work (if you make it to page 4 or back to index it is disabled)  but not from the get go. 

Also some issues with page. it doesn't cycle through cleanly. (Im doing some stupid stuff. will look through your earlier examples tomorrow)

am i right in thinking the second batch of code , replaces the earlier stuff? 
many thanks for getting me through this. am slowly getting it.

**Comment by harrydry** [comment]
scrap the above, your code works perfect. 
will confirm tomorrow. (another thing i got to test) 

many thanks

**Comment by harrydry** [comment]
still a slight problem

**Comment by luruke** [comment]
@harrydry I'll close this.

If you have any other problem open a new issue and be more precise.
Cheers.
